engines/gnap/gamesys.cpp

```diff
@@ -0,0 +1,1278 @@
+			} else if (gfxItem5->_flags & 2) {
+				/* TODO
+				spriteData4 = gfxItem5->prevFrame.spriteId;
+				for (int n = 0; n < gfxItem5->updRectsCount; ++n)
+					seqDrawAviFrame(spriteData4, &gfxItem5->prevFrame, &gfxItem5->updRects[n], gfxItem5->flags & 8);
+				*/
+			} else {
+				int resourceId = (gfxItem5->_sequenceId & 0xFFFF0000) | gfxItem5->_prevFrame._spriteId;
+				SpriteResource *spriteResource = _vm->_spriteCache->get(resourceId);
+				for (int n = 0; n < gfxItem5->_updRectsCount; ++n)
+					seqDrawSpriteFrame(spriteResource, gfxItem5->_prevFrame, &gfxItem5->_updRects[n]);
+				_vm->_spriteCache->release(resourceId);
+			}
+		}
+	}
+
+	debugC(kDebugBasic, "GameSys::drawSprites() OK");
+}
+
+void GameSys::updateRect(const Common::Rect &r) {
+	debugC(kDebugBasic, "GameSys::updateRect() %d, %d, %d, %d [%d, %d]", r.left, r.top, r.right, r.bottom, r.width(), r.height());
+	if (r.width() > 0 && r.height() > 0) {
+		byte *pixels = (byte*)_frontSurface->getBasePtr(r.left, r.top);
+		_vm->_system->copyRectToScreen(pixels, _frontSurface->pitch, r.left, r.top,
+			r.width(), r.height());
+	}
+}
+
+void GameSys::updateScreen() {
+	debugC(kDebugBasic, "GameSys::updateScreen()");
+
+	for (uint i = 0; i < _dirtyRects.size(); ++i)
+		updateRect(_dirtyRects[i]);
+
+	if (_dirtyRects.size() > 0) {
+		_dirtyRects.clear();
+		_lastUpdateClock = 0;
+		_gameSysClock = 0;
+	}
+
+	Common::Rect dstRect, srcRect, rcSrc2;
+
+	for (int j = 0; j < _gfxItemsCount; ++j) {
+
+		GfxItem *gfxItem = &_gfxItems[j];
+
+		if (!gfxItem->_updFlag)
+			continue;
+
+		if (gfxItem->_prevFrame._spriteId == -1 ||
+			!intersectRect(srcRect, _screenRect, gfxItem->_prevFrame._rect)) {
+			if (gfxItem->_currFrame._spriteId != -1 && intersectRect(rcSrc2, _screenRect, gfxItem->_currFrame._rect))
+				updateRect(rcSrc2);
+		} else if (gfxItem->_currFrame._spriteId != -1 &&
+			intersectRect(rcSrc2, _screenRect, gfxItem->_currFrame._rect)) {
+			updateRect(srcRect);
+			updateRect(rcSrc2);
+		}
+		gfxItem->_prevFrame = gfxItem->_currFrame;
+	}
+
+	updateRect(Common::Rect(0, 0, 800, 600));
+
+	debugC(kDebugBasic, "GameSys::updateScreen() OK");
+}
+
+void GameSys::handleReqRemoveSequenceItem() {
+	if (_reqRemoveSequenceItem) {
+		int gfxIndex2;
+		_reqRemoveSequenceItem = false;
+		if (seqFind(_removeSequenceItemSequenceId, _removeSequenceItemValue, &gfxIndex2))
+			_seqItems.remove_at(gfxIndex2);
+		if (seqLocateGfx(_removeSequenceItemSequenceId, _removeSequenceItemValue, &gfxIndex2)) {
+			int gfxIndex2a = gfxIndex2;
+			for (GfxItem *gfxItem = &_gfxItems[gfxIndex2a];
+				gfxIndex2a < _gfxItemsCount && gfxItem->_sequenceId == _removeSequenceItemSequenceId && gfxItem->_id == _removeSequenceItemValue;
+				gfxItem = &_gfxItems[gfxIndex2a])
+				++gfxIndex2a;
+			_gfxItemsCount -= gfxIndex2a - gfxIndex2;
+			if (_gfxItemsCount != gfxIndex2)
+				memcpy(&_gfxItems[gfxIndex2], &_gfxItems[gfxIndex2a], sizeof(GfxItem) * (_gfxItemsCount - gfxIndex2));
```

Comment:
I didn't get any Valgrind warning about overlapping memory areas for this one, but it looks to me like it could happen here as well, so we should use memmove() instead.